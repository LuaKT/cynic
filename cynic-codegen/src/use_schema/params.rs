#[derive(Debug)]
pub struct UseSchemaParams {
    pub schema_filename: String,
}

impl UseSchemaParams {
    fn new(schema_filename: String) -> Self {
        UseSchemaParams { schema_filename }
    }
}

impl syn::parse::Parse for UseSchemaParams {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let lit_str = input.parse::<syn::LitStr>()?;

        Ok(UseSchemaParams {
            schema_filename: lit_str.value(),
        })
    }
}
