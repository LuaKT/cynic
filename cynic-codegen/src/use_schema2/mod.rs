//mod model;
mod named_type;
mod params;
mod schema_roots;
mod subtype_markers;

pub use params::UseSchemaParams;

use proc_macro2::TokenStream;
use quote::format_ident;
use syn::parse_quote;

use crate::{
    field_type::FieldType,
    ident::{to_snake_case, Ident},
    schema::{self, SchemaLoadError},
    type_index::TypeIndex,
};

use self::{named_type::NamedType, subtype_markers::SubtypeMarkers};

pub fn use_schema(input: UseSchemaParams) -> Result<TokenStream, SchemaLoadError> {
    use quote::{quote, TokenStreamExt};

    let document = crate::schema::load_schema(input.schema_filename)?;

    let mut output = TokenStream::new();

    let root_types = schema_roots::RootTypes::from_definitions(&document.definitions);
    output.append_all(quote! {
        #root_types
    });

    let mut subtype_markers = Vec::new();
    let mut named_types = Vec::new();

    // TODO: Refactor this so it's not just one big loop
    for definition in document
        .definitions
        .into_iter()
        .filter_map(type_def_from_definition)
    {
        named_types.extend(NamedType::from_def(&definition));

        match definition {
            graphql_parser::schema::TypeDefinition::Scalar(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            graphql_parser::schema::TypeDefinition::Object(def) => {
                subtype_markers.extend(SubtypeMarkers::from_object(&def));

                let object_marker_type_name = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #object_marker_type_name;
                });

                let mut field_module_contents = Vec::new();
                for field in def.fields {
                    // TODO: wonder if we need a different Ident func for this.
                    // Technically field.name is a field, but we're only using it
                    // as a type.
                    let field_marker_type_name = Ident::for_type(&field.name);
                    let field_name_literal = proc_macro2::Literal::string(&field.name);

                    // Note: See if we can get rid of TypeIndex from this
                    // call once done.  That'd be nice.
                    let field_type =
                        FieldType::from_schema_type(&field.field_type, &TypeIndex::empty());
                    let field_type_marker = field_type.to_tokens(None, parse_quote! { super });

                    field_module_contents.push(quote! {
                        pub struct #field_marker_type_name;

                        impl ::cynic::schema::Field for #field_marker_type_name {
                            type SchemaType = #field_type_marker;

                            fn name() -> &'static str {
                                #field_name_literal
                            }
                        }

                        impl ::cynic::schema::HasField<#field_marker_type_name, #field_type_marker> for super::#object_marker_type_name {}

                        // TODO: implement HasField for all the valid conversions...
                        // assuming that's even possible - implementing the deserialize might be tricky for
                        // some of them?  Need to check what's even allowed here...
                    })

                    // TODO: Handle arguments
                }

                let field_module_name = format_ident!("{}_fields", to_snake_case(&def.name));
                output.append_all(quote! {
                    pub mod #field_module_name {
                        #(#field_module_contents)*
                    }
                });
            }
            graphql_parser::schema::TypeDefinition::Interface(def) => {
                subtype_markers.push(SubtypeMarkers::from_interface(&def));

                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: the rest of this.  Presumably we need fields & HasSubtype
            }
            graphql_parser::schema::TypeDefinition::Union(def) => {
                subtype_markers.extend(SubtypeMarkers::from_union(&def));

                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: the rest of this.  Presumably we need just HasSubtype
            }
            graphql_parser::schema::TypeDefinition::Enum(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });
            }
            graphql_parser::schema::TypeDefinition::InputObject(def) => {
                let ident = Ident::for_type(&def.name);
                output.append_all(quote! {
                    pub struct #ident {}
                });

                // TODO: Handle fields
            }
        }
    }

    output.append_all(quote! {
        #(#subtype_markers)*
        #(#named_types)*
    });

    // TODO: output defs for the built in scalars so we don't have
    // to special case them elsewhere...

    Ok(output)
}

fn type_def_from_definition(definition: schema::Definition) -> Option<schema::TypeDefinition> {
    match definition {
        graphql_parser::schema::Definition::TypeDefinition(inner) => Some(inner),
        _ => None,
    }
}
