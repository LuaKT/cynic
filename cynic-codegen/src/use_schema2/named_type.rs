use proc_macro2::TokenStream;

use crate::{idents::Ident, schema::TypeDefinition};

pub struct NamedType {
    ident: super::Ident,
}

impl NamedType {
    pub fn from_def(def: &TypeDefinition) -> Option<Self> {
        match def {
            // Note: Currently we only use the NamedType lookup for members
            // of interfaces & unions - so we specifically don't generate anything for
            // scalars, inputs or enums.
            TypeDefinition::Scalar(_) => None,
            TypeDefinition::InputObject(_) => None,
            TypeDefinition::Enum(_) => None,

            TypeDefinition::Object(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
            TypeDefinition::Interface(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
            TypeDefinition::Union(def) => Some(NamedType {
                ident: Ident::for_type(&def.name),
            }),
        }
    }
}

impl quote::ToTokens for NamedType {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let rust_name = &self.ident;
        let graphql_name = proc_macro2::Literal::string(self.ident.graphql_name());

        tokens.append_all(quote! {
            impl ::cynic::schema::NamedType for #rust_name {
                fn name() -> &'static str {
                    #graphql_name
                }
            }
        });
    }
}
