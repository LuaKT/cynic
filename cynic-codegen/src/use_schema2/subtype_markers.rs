use proc_macro2::TokenStream;

use crate::{schema, Ident};

/// Outputs `HasSubtype` implementations for the given types.
#[derive(Debug)]
pub struct SubtypeMarkers {
    pub subtype: Ident,
    pub supertypes: Vec<Ident>,
}

impl SubtypeMarkers {
    pub fn from_interface(iface: &schema::InterfaceType) -> Self {
        let ident = Ident::for_type(&iface.name);

        Self {
            subtype: ident.clone(),
            supertypes: vec![ident],
        }
    }

    pub fn from_union(def: &schema::UnionType) -> Vec<Self> {
        let supertype = Ident::for_type(&def.name);

        def.types
            .iter()
            .map(|ty| SubtypeMarkers {
                subtype: Ident::for_type(ty),
                supertypes: vec![supertype.clone()],
            })
            .collect()
    }

    pub fn from_object(obj: &schema::ObjectType) -> Option<Self> {
        if obj.implements_interfaces.is_empty() {
            return None;
        }

        Some(Self {
            subtype: Ident::for_type(&obj.name),
            supertypes: obj
                .implements_interfaces
                .iter()
                .map(Ident::for_type)
                .collect(),
        })
    }
}

impl quote::ToTokens for SubtypeMarkers {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let subtype = &self.subtype;
        let supertypes = &self.supertypes;

        tokens.append_all(quote! {
            #(
                impl ::cynic::schema::HasSubtype<#subtype> for #supertypes {}
            )*
        });
    }
}
