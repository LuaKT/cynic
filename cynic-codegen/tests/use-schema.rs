use std::{io::Write, path::PathBuf, process::Stdio};

use insta::assert_snapshot;
use rstest::rstest;

use cynic_codegen::use_schema::{use_schema, QueryDslParams};
use cynic_codegen::use_schema2::{use_schema as use_schema_2, UseSchemaParams as UseSchema2Params};

// TODO: Rename this file after running snapshots

#[rstest(schema_file => [
    "graphql.jobs.graphql",
    "books.graphql",
    "starwars.schema.graphql",
    "test_cases.graphql",
])]
fn snapshot_use_schema(schema_file: &str) {
    let schema_path = PathBuf::from("../schemas/").join(schema_file);

    let tokens = use_schema(QueryDslParams {
        schema_filename: schema_path.to_str().unwrap().to_string(),
    })
    .unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

#[rstest(schema_file_two => [
    "graphql.jobs.graphql",
    "books.graphql",
    "starwars.schema.graphql",
    "test_cases.graphql",
    "../cynic/src/bin/simple.graphql"
])]
fn snapshot_use_schema_two(schema_file_two: &str) {
    let schema_path = PathBuf::from("../schemas/").join(schema_file_two);

    let tokens = use_schema_2(UseSchema2Params {
        schema_filename: schema_path.to_str().unwrap().to_string(),
    })
    .unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

fn format_code(input: String) -> String {
    let mut cmd = std::process::Command::new("rustfmt")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .expect("failed to execute rustfmt");

    write!(cmd.stdin.as_mut().unwrap(), "{}", input).unwrap();

    std::str::from_utf8(&cmd.wait_with_output().unwrap().stdout)
        .unwrap()
        .to_owned()
}
