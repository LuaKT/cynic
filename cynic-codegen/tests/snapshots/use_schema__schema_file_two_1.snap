---
source: cynic-codegen/tests/use-schema.rs
expression: "format_code(format!(\"{}\", tokens))"

---
pub struct City;
pub mod city_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::City {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::City {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::City {}
    pub struct Country;
    impl ::cynic::core::FieldName for Country {
        fn name() -> &'static str {
            "country"
        }
    }
    impl ::cynic::core::HasField<Country, super::Country> for super::City {}
    pub struct Type;
    impl ::cynic::core::FieldName for Type {
        fn name() -> &'static str {
            "type"
        }
    }
    impl ::cynic::core::HasField<Type, String> for super::City {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::City {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::City {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::City {}
}
pub struct CityOrderByInput {}
pub struct CityWhereInput {}
pub struct Commitment;
pub mod commitment_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Commitment {}
    pub struct Title;
    impl ::cynic::core::FieldName for Title {
        fn name() -> &'static str {
            "title"
        }
    }
    impl ::cynic::core::HasField<Title, String> for super::Commitment {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Commitment {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::Commitment {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Commitment {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Commitment {}
}
pub struct CommitmentWhereInput {}
pub struct Company;
pub mod company_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Company {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::Company {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Company {}
    pub struct WebsiteUrl;
    impl ::cynic::core::FieldName for WebsiteUrl {
        fn name() -> &'static str {
            "websiteUrl"
        }
    }
    impl ::cynic::core::HasField<WebsiteUrl, String> for super::Company {}
    pub struct LogoUrl;
    impl ::cynic::core::FieldName for LogoUrl {
        fn name() -> &'static str {
            "logoUrl"
        }
    }
    impl ::cynic::core::HasField<LogoUrl, Option<String>> for super::Company {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::Company {}
    pub struct Twitter;
    impl ::cynic::core::FieldName for Twitter {
        fn name() -> &'static str {
            "twitter"
        }
    }
    impl ::cynic::core::HasField<Twitter, Option<String>> for super::Company {}
    pub struct Emailed;
    impl ::cynic::core::FieldName for Emailed {
        fn name() -> &'static str {
            "emailed"
        }
    }
    impl ::cynic::core::HasField<Emailed, Option<bool>> for super::Company {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Company {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Company {}
}
pub struct CompanyWhereInput {}
pub struct Country;
pub mod country_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Country {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::Country {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Country {}
    pub struct Type;
    impl ::cynic::core::FieldName for Type {
        fn name() -> &'static str {
            "type"
        }
    }
    impl ::cynic::core::HasField<Type, String> for super::Country {}
    pub struct IsoCode;
    impl ::cynic::core::FieldName for IsoCode {
        fn name() -> &'static str {
            "isoCode"
        }
    }
    impl ::cynic::core::HasField<IsoCode, Option<String>> for super::Country {}
    pub struct Cities;
    impl ::cynic::core::FieldName for Cities {
        fn name() -> &'static str {
            "cities"
        }
    }
    impl ::cynic::core::HasField<Cities, Option<Vec<super::City>>> for super::Country {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::Country {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Country {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Country {}
}
pub struct CountryOrderByInput {}
pub struct CountryWhereInput {}
pub struct DateTime {}
pub struct Job;
pub mod job_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Job {}
    pub struct Title;
    impl ::cynic::core::FieldName for Title {
        fn name() -> &'static str {
            "title"
        }
    }
    impl ::cynic::core::HasField<Title, String> for super::Job {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Job {}
    pub struct Commitment;
    impl ::cynic::core::FieldName for Commitment {
        fn name() -> &'static str {
            "commitment"
        }
    }
    impl ::cynic::core::HasField<Commitment, super::Commitment> for super::Job {}
    pub struct Cities;
    impl ::cynic::core::FieldName for Cities {
        fn name() -> &'static str {
            "cities"
        }
    }
    impl ::cynic::core::HasField<Cities, Option<Vec<super::City>>> for super::Job {}
    pub struct Countries;
    impl ::cynic::core::FieldName for Countries {
        fn name() -> &'static str {
            "countries"
        }
    }
    impl ::cynic::core::HasField<Countries, Option<Vec<super::Country>>> for super::Job {}
    pub struct Remotes;
    impl ::cynic::core::FieldName for Remotes {
        fn name() -> &'static str {
            "remotes"
        }
    }
    impl ::cynic::core::HasField<Remotes, Option<Vec<super::Remote>>> for super::Job {}
    pub struct Description;
    impl ::cynic::core::FieldName for Description {
        fn name() -> &'static str {
            "description"
        }
    }
    impl ::cynic::core::HasField<Description, Option<String>> for super::Job {}
    pub struct ApplyUrl;
    impl ::cynic::core::FieldName for ApplyUrl {
        fn name() -> &'static str {
            "applyUrl"
        }
    }
    impl ::cynic::core::HasField<ApplyUrl, Option<String>> for super::Job {}
    pub struct Company;
    impl ::cynic::core::FieldName for Company {
        fn name() -> &'static str {
            "company"
        }
    }
    impl ::cynic::core::HasField<Company, Option<super::Company>> for super::Job {}
    pub struct Tags;
    impl ::cynic::core::FieldName for Tags {
        fn name() -> &'static str {
            "tags"
        }
    }
    impl ::cynic::core::HasField<Tags, Option<Vec<super::Tag>>> for super::Job {}
    pub struct IsPublished;
    impl ::cynic::core::FieldName for IsPublished {
        fn name() -> &'static str {
            "isPublished"
        }
    }
    impl ::cynic::core::HasField<IsPublished, Option<bool>> for super::Job {}
    pub struct IsFeatured;
    impl ::cynic::core::FieldName for IsFeatured {
        fn name() -> &'static str {
            "isFeatured"
        }
    }
    impl ::cynic::core::HasField<IsFeatured, Option<bool>> for super::Job {}
    pub struct LocationNames;
    impl ::cynic::core::FieldName for LocationNames {
        fn name() -> &'static str {
            "locationNames"
        }
    }
    impl ::cynic::core::HasField<LocationNames, Option<String>> for super::Job {}
    pub struct UserEmail;
    impl ::cynic::core::FieldName for UserEmail {
        fn name() -> &'static str {
            "userEmail"
        }
    }
    impl ::cynic::core::HasField<UserEmail, Option<String>> for super::Job {}
    pub struct PostedAt;
    impl ::cynic::core::FieldName for PostedAt {
        fn name() -> &'static str {
            "postedAt"
        }
    }
    impl ::cynic::core::HasField<PostedAt, super::DateTime> for super::Job {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Job {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Job {}
}
pub struct JobInput {}
pub struct JobOrderByInput {}
pub struct JobsInput {}
pub struct JobWhereInput {}
pub struct Location;
pub mod location_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Location {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Location {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::Location {}
    pub struct Type;
    impl ::cynic::core::FieldName for Type {
        fn name() -> &'static str {
            "type"
        }
    }
    impl ::cynic::core::HasField<Type, String> for super::Location {}
}
pub struct LocationInput {}
pub struct LocationsInput {}
pub struct Mutation;
pub mod mutation_fields {
    pub struct Subscribe;
    impl ::cynic::core::FieldName for Subscribe {
        fn name() -> &'static str {
            "subscribe"
        }
    }
    impl ::cynic::core::HasField<Subscribe, super::User> for super::Mutation {}
    pub struct PostJob;
    impl ::cynic::core::FieldName for PostJob {
        fn name() -> &'static str {
            "postJob"
        }
    }
    impl ::cynic::core::HasField<PostJob, super::Job> for super::Mutation {}
    pub struct UpdateJob;
    impl ::cynic::core::FieldName for UpdateJob {
        fn name() -> &'static str {
            "updateJob"
        }
    }
    impl ::cynic::core::HasField<UpdateJob, super::Job> for super::Mutation {}
    pub struct UpdateCompany;
    impl ::cynic::core::FieldName for UpdateCompany {
        fn name() -> &'static str {
            "updateCompany"
        }
    }
    impl ::cynic::core::HasField<UpdateCompany, super::Company> for super::Mutation {}
}
pub struct PostJobInput {}
pub struct Query;
pub mod query_fields {
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Vec<super::Job>> for super::Query {}
    pub struct Job;
    impl ::cynic::core::FieldName for Job {
        fn name() -> &'static str {
            "job"
        }
    }
    impl ::cynic::core::HasField<Job, super::Job> for super::Query {}
    pub struct Locations;
    impl ::cynic::core::FieldName for Locations {
        fn name() -> &'static str {
            "locations"
        }
    }
    impl ::cynic::core::HasField<Locations, Vec<super::Location>> for super::Query {}
    pub struct City;
    impl ::cynic::core::FieldName for City {
        fn name() -> &'static str {
            "city"
        }
    }
    impl ::cynic::core::HasField<City, super::City> for super::Query {}
    pub struct Country;
    impl ::cynic::core::FieldName for Country {
        fn name() -> &'static str {
            "country"
        }
    }
    impl ::cynic::core::HasField<Country, super::Country> for super::Query {}
    pub struct Remote;
    impl ::cynic::core::FieldName for Remote {
        fn name() -> &'static str {
            "remote"
        }
    }
    impl ::cynic::core::HasField<Remote, super::Remote> for super::Query {}
    pub struct Commitments;
    impl ::cynic::core::FieldName for Commitments {
        fn name() -> &'static str {
            "commitments"
        }
    }
    impl ::cynic::core::HasField<Commitments, Vec<super::Commitment>> for super::Query {}
    pub struct Cities;
    impl ::cynic::core::FieldName for Cities {
        fn name() -> &'static str {
            "cities"
        }
    }
    impl ::cynic::core::HasField<Cities, Vec<super::City>> for super::Query {}
    pub struct Countries;
    impl ::cynic::core::FieldName for Countries {
        fn name() -> &'static str {
            "countries"
        }
    }
    impl ::cynic::core::HasField<Countries, Vec<super::Country>> for super::Query {}
    pub struct Remotes;
    impl ::cynic::core::FieldName for Remotes {
        fn name() -> &'static str {
            "remotes"
        }
    }
    impl ::cynic::core::HasField<Remotes, Vec<super::Remote>> for super::Query {}
    pub struct Companies;
    impl ::cynic::core::FieldName for Companies {
        fn name() -> &'static str {
            "companies"
        }
    }
    impl ::cynic::core::HasField<Companies, Vec<super::Company>> for super::Query {}
}
pub struct Remote;
pub mod remote_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Remote {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::Remote {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Remote {}
    pub struct Type;
    impl ::cynic::core::FieldName for Type {
        fn name() -> &'static str {
            "type"
        }
    }
    impl ::cynic::core::HasField<Type, String> for super::Remote {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::Remote {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Remote {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Remote {}
}
pub struct RemoteOrderByInput {}
pub struct RemoteWhereInput {}
pub struct SubscribeInput {}
pub struct Tag;
pub mod tag_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::Tag {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, String> for super::Tag {}
    pub struct Slug;
    impl ::cynic::core::FieldName for Slug {
        fn name() -> &'static str {
            "slug"
        }
    }
    impl ::cynic::core::HasField<Slug, String> for super::Tag {}
    pub struct Jobs;
    impl ::cynic::core::FieldName for Jobs {
        fn name() -> &'static str {
            "jobs"
        }
    }
    impl ::cynic::core::HasField<Jobs, Option<Vec<super::Job>>> for super::Tag {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::Tag {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::Tag {}
}
pub struct TagOrderByInput {}
pub struct TagWhereInput {}
pub struct UpdateCompanyInput {}
pub struct UpdateJobInput {}
pub struct User;
pub mod user_fields {
    pub struct Id;
    impl ::cynic::core::FieldName for Id {
        fn name() -> &'static str {
            "id"
        }
    }
    impl ::cynic::core::HasField<Id, ::cynic::Id> for super::User {}
    pub struct Name;
    impl ::cynic::core::FieldName for Name {
        fn name() -> &'static str {
            "name"
        }
    }
    impl ::cynic::core::HasField<Name, Option<String>> for super::User {}
    pub struct Email;
    impl ::cynic::core::FieldName for Email {
        fn name() -> &'static str {
            "email"
        }
    }
    impl ::cynic::core::HasField<Email, String> for super::User {}
    pub struct Subscribe;
    impl ::cynic::core::FieldName for Subscribe {
        fn name() -> &'static str {
            "subscribe"
        }
    }
    impl ::cynic::core::HasField<Subscribe, bool> for super::User {}
    pub struct CreatedAt;
    impl ::cynic::core::FieldName for CreatedAt {
        fn name() -> &'static str {
            "createdAt"
        }
    }
    impl ::cynic::core::HasField<CreatedAt, super::DateTime> for super::User {}
    pub struct UpdatedAt;
    impl ::cynic::core::FieldName for UpdatedAt {
        fn name() -> &'static str {
            "updatedAt"
        }
    }
    impl ::cynic::core::HasField<UpdatedAt, super::DateTime> for super::User {}
}

